src/rules/maxMethodFunctionLineCountRule.ts

```diff
@@ -0,0 +1,108 @@
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "max-method-function-line-count",
+        description: "Requires methods or functions to remain under a certain number of lines",
+        rationale: Lint.Utils.dedent`
+            Limiting the number of lines allowed in a block allows blocks to remain small,
+            single purpose, and maintainable.`,
+        optionsDescription: Lint.Utils.dedent`
+            An integer indicating the maximum line count of functions and methods.
+            An optional boolean indicating if nested functions of functions are included in the count (default true).`,
+        options: {
+            type: "array",
+            items: [
+                {
+                    type: "number",
+                },
+                {
+                    type: "boolean",
+                },
+            ],
+            additionalItems: false,
+        },
+        optionExamples: [[true, 200, false]],
+        type: "maintainability",
+        typescriptOnly: false,
+    };
+
+    public isEnabled(): boolean {
+        return super.isEnabled() && this.ruleArguments[0] as number > 0;
+    }
+
+    public includesNested(): boolean {
+        if (this.ruleArguments.length > 1) {
+            return this.ruleArguments[1] as boolean;
+        }
+        return true;
+    }
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new MaxMethodLine(sourceFile, this.ruleName, {
+            limit: this.ruleArguments[0] as number,
+            includesNested: this.includesNested(),
+        }));
+    }
+}
+
+function FAILURE_STRING(lineCount: number, lineLimit: number) {
+    return `This method or function has ${lineCount} lines, which exceeds the maximum of ${lineLimit} lines allowed. ` +
+        "Consider breaking this up into smaller parts.";
+}
+
+class MaxMethodLine extends Lint.AbstractWalker<{limit: number; includesNested: boolean}> {
+    public walk(sourceFile: ts.SourceFile) {
+        const cb = (node: ts.Node): void => {
+            if (isFunctionWithBody(node)) {
+                const limit = this.options.limit;
+                const lines = this.countLines(node.body!);
+                if (lines > limit) {
+                    this.addFailure(node.getStart(this.sourceFile), node.body!.pos, FAILURE_STRING(lines, limit));
+                }
+            }
+            return ts.forEachChild(node, cb);
+        };
+        return ts.forEachChild(sourceFile, cb);
+    }
+
+    private countLines(node: ts.Node) {
+        const includesNested = this.options.includesNested;
+        let end = node.end;
+        if (!includesNested) {
+            const firstNestedFunction = (anode: ts.Node): ts.Node | undefined => {
+                if (isFunctionWithBody(anode)) {
+                    return anode;
+                }
+                return ts.forEachChild(anode, firstNestedFunction);
```

Comment:
this approach fails for code like this:
```ts
function reallyLong() {
    let foo = function() {}; // you end here and consider this line the end of the function
    statement;
    statement;
    function foo() {}
    statement;
    // ... and so on
}
```

Instead you probably want to find the last statement that is not a function declaration:

```ts
// node: ts.FunctionLikeDeclaration
// node.body: ts.Block
for (let i = node.body.statements.length - 1; i >= 0; --i) {
    if (!isFunctionDeclaration(node.body.statements[i])) {
        break;
    }
    end = node.body.statements[i].end;
}

```

Reply:
@ajafff I have troubles to check if my node is a ts.FunctionLikeDeclaration because I don't have isFunctionLikeDeclaration. I want to perform this test to cast my node.

Is it normal that isFunctionLikeDeclaration is missing ?

Reply:
Yeah, seems like I never implemented that typeguard in tsutils. Anyways, I suggested `isFunctionDeclaration` (without the `Like`) because that's the only declaration that's also a statement.
Every other functionlike declaration is an expression, e.g. ArrowFunction and FunctionExpression, and will therefore never be true when checking statements: `isFunctionExpression(block.statements[i])` is always false